pkg/gadget-service/api/api.proto

```diff
@@ -72,6 +72,14 @@ message GadgetRunRequest {
   int64 timeout = 13;
 }
 
+message GadgetAttachRequest {
+  // id of the gadget to attach to
+  string id = 1;
+
+  // used to inform the server about the expected protocol version
+  uint32 version = 2;
```

Comment:
It's not versioning the gRPC protocol nor the message layout by itself, but rather a generic version that needs to be agreed on between client & server. I think it's way easier to communicate that than inferring it from present / not-present fields and whether they're filled or not filled.

Currently we hold those version numbers for `GetGadgetInfo` and `RunGadget` [here](https://github.com/inspektor-gadget/inspektor-gadget/blob/30fd9315357eef371fab9f68fcc2491f2553fd1c/pkg/gadget-service/api/consts.go#L18).

> For example, if we need to add a parameter uint32 terminalWidth = 3;, we can just do that and the server can just test if it has the default value (0) to know if the client has set the field.

Not sure I can follow you there. `0` could be a valid value for `terminalWidth` - but as it is also the default value for ints, afaik in protobuf it doesn't get sent over the wire. So it's not an indication of whether the client was aware of that field and infer a "version" from it. (I'll re-check the docs, though)

Again, happy to document once we agreed on the internals.

Reply:
> Not sure I can follow you there. 0 could be a valid value for terminalWidth - but as it is also the default value for ints, afaik in protobuf it doesn't get sent over the wire. So it's not an indication of whether the client was aware of that field and infer a "version" from it. (I'll re-check the docs, though)

That's true, that can only be done when 0 is not a valid value for `terminalWidth`. When adding a field where the default value (0 for integer) is a valid value, this can be done with making a new method with a suffix such as V2 taking a parameter GadgetAttachRequestV2 as a suffix. Or at the package layer ([example in Kubernetes cri-api](https://github.com/kubernetes/cri-api/blob/010fdf8d785a7f35fa5a1cc80c40bde0c13e1dfd/pkg/apis/runtime/v1/api.proto#L20)).

My issue with the `version` field is that it is not self-documenting, as opposed to the other 3 methods above (1. implicit when the new field default value is invalid, 2. new methods/structs with a V2 suffix, 3. version at the package level).

If you still prefer the `version` field, the comment should be more clear in this PR: list the acceptable values, and link to the constant definition and to the code checking the value.

Reply:
I don't have a super strong opinion on how to handle this. The version field right now can communicate incompatibilities on a logical level without the need for incompatible proto messages / rpcs. The 0 as default being valid or not is something that I'd rather not rely on, but versioning using specific packages or suffixes is probably a good alternative (although I don't know what the UX will be like in incompatible situations and how much control you've got over it - but it's worth checking out).

Anyway, since we're still using this version field for the existing RPCs, I'd like to move forward the way it is right now and move the discussion to after merging this PR. Having a version field now shouldn't prevent us from going one of the routes you suggested.